core/src/services/alluxio/writer.rs

```diff
@@ -16,42 +16,177 @@
+                        }));
+                    }
+                },
+                State::Init(fut) => {
+                    let (w, stream_id) = ready!(fut.as_mut().poll(cx));
+                    self.state = State::Idle(Some(w));
+                    self.stream_id = Some(stream_id?);
+                }
+                State::Write(fut) => {
+                    let (w, part) = ready!(fut.as_mut().poll(cx));
+                    self.state = State::Idle(Some(w));
+                    return Poll::Ready(Ok(part?));
+                }
+                State::Close(_) => {
+                    unreachable!(
+                        "MultipartUploadWriter must not go into State::Close during poll_write"
+                    )
+                }
+                State::Abort(_) => {
+                    unreachable!(
+                        "MultipartUploadWriter must not go into State::Abort during poll_write"
+                    )
+                }
+            }
+        }
+    }
+
+    fn poll_close(&mut self, cx: &mut Context<'_>) -> Poll<Result<()>> {
+        loop {
+            match &mut self.state {
+                State::Idle(w) => {
+                    let w = w.take().expect("writer must be valid");
+                    match self.stream_id {
+                        Some(stream_id) => {
+                            self.state = State::Close(Box::pin(async move {
+                                let res = w.close(stream_id).await;
+                                (w, res)
+                            }));
+                        }
+                        None => {
+                            return Poll::Ready(Ok(()));
+                        }
+                    }
+                }
+                State::Close(fut) => {
+                    let (w, res) = futures::ready!(fut.as_mut().poll(cx));
+                    self.state = State::Idle(Some(w));
+
+                    res?;
+
+                    return Poll::Ready(Ok(()));
+                }
+                State::Init(_) => {
+                    unreachable!("AlluxioWriter must not go into State::Init during poll_close")
+                }
+                State::Write(_) => unreachable! {
+                    "AlluxioWriter must not go into State::Write during poll_close"
+                },
+                State::Abort(_) => {
+                    unreachable!("AlluxioWriter must not go into State::Abort during poll_close")
+                }
+            }
+        }
+    }
 
-        Ok(())
+    fn poll_abort(&mut self, cx: &mut Context<'_>) -> Poll<Result<()>> {
+        loop {
+            match &mut self.state {
+                State::Idle(w) => {
+                    let w = w.take().expect("writer must be valid");
+                    match self.stream_id {
+                        Some(_) => {
+                            let path = self.path.clone();
+                            self.state = State::Abort(Box::pin(async move {
```

Comment:
Please don't implement abort by calling `delete`. Leave this as unsupported if service doesn't provide native abort support. We can remove `Abort` state.

Reply:
fixed